packages/typespec-ts/src/utils/modelUtils.ts

```diff
@@ -749,7 +760,14 @@ function getSchemaForModel(
       name === `"${propertyName}"` &&
       modelSchema.discriminator
     ) {
-      modelSchema.discriminator.type = propSchema.typeName ?? propSchema.type;
+      modelSchema.discriminator = {
+        ...modelSchema.discriminator,
+        ...{
+          type: propSchema.typeName ?? propSchema.type,
+          typeName: propSchema.typeName,
+          outputTypeName: propSchema.outputTypeName
```

Comment:
with union as discriminator, we have to know its typeName and outputTypeName when generating the code.